### DISCUSSION TITLE
Info Feature oder Info Widget

### DISCUSSION THREAD

**Discussion by superfunk2000** [Lissy93/dashy#231] [Ideas]
Hello Lissy93,
thank you very much for this great dashy! I replaced Homer with it and I am very satisfied.

Widgets are planned soon, is there an info widget included?
How I imagine it exactly:
If a certain text file exists, its content is displayed as popup-info on Dashy. After confirming the information, the file will be deleted. In addition to the information, a heading and a (Font Awesom) icon can also be specified in the content of the text file.
Maybe like this:
```
wdgInfo:
  icon: "fas fa-water"
  header: "Warning"
  info: "Water sensor pump has triggered" 
```
Greetings from Germany
Superfunk2000

**Comment by superfunk2000** [comment]
The confirm should be able to be deactivated. The info disappears when the file disappears. (Adjustable verification time)
Furthermore, you should be able to decide whether it is a popup info or a static one that is displayed above. Maybe with an adjustable background color.

**Comment by lissy93** [comment]
Thinking about it, a better implementation may be to use either integrate with a system like [Gotify](https://gotify.net/), or to read notifications via an existing protocol, such as RSS. That way events can show up immediately, be dismissed through the UI, and be easily hooked up to other apps and services.
I think that reading and writing files to disk frequently may add more complexity than it's worth, especially in terms of security and efficiently monitoring changes to a filesystem.

**Comment by superfunk2000** [comment]
Yes, using Gotify would also be a good and simple solution.
Do you already have a guess when the widgets will come?

**Reply by lissy93** [reply]
Planned for V 2.0.0, and we're on V 1.8.1 at the moment. Most likley 4 - 6 weeks :)

**Reply by lissy93** [reply]
Just a quick update on this - I'm working on widgets at the moment. There's an open PR (#382) which should be merged sometime within the next week with alpha support for widgets. It's early days still, buy I've made a few pre-built widgets, which [you can see here](https://github.com/Lissy93/dashy/blob/FEATURE/basic-widget-support/docs/widgets.md). Let me know what you think :)

<p align="center">
  <img width="600" src="https://i.ibb.co/GFjXVHy/dashy-widgets.png" />
</p>

**Reply by lissy93** [reply]
Merged and released in V 1.9.4. See the [Widget Docs](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md) for info on getting started :)

**Reply by EVOTk** [reply]
> Merged and released in V 1.9.4. See the [Widget Docs](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md) for info on getting started :)

I added PiHole, but it doesn't work, I must have missed something!

How to see the logs? Thank you

```
  - name: Serveur DNS
    widgets:
      - type: pi-hole-stats
        options:
          hostname: http://192.168.2.90
```

![2021-12-29 20_46_24-Window](https://user-images.githubusercontent.com/45015615/147698835-e37701a6-9ead-48ac-abf4-75cbe9feb8aa.png)

**Reply by lissy93** [reply]
Would you mind sharing the output in the browser console?
Your config is correct ✅. Maybe I need to add an option for authorization headers :thinking:

**Reply by EVOTk** [reply]
Thank you for your reply ;)

```
  - name: Serveur DNS
    widgets:
      - type: pi-hole-stats
        options:
          hostname: http://192.168.2.90
      - type: pi-hole-top-queries
        options:
          hostname: http://192.168.2.90
          apiKey: MY_API_KEY
```

I also tried "pi-hole-top-queries" which gives the same result, here is what I find in the console:

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://192.168.2.90/admin/api.php?topItems=10&auth=MY_API_KEY. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://192.168.2.90/admin/api.php. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.

**Reply by lissy93** [reply]
Ah, that's a CORS error :(
I think that I can fix this by adding an option to proxy requests through the Node server

Are you serving Pi-Hole through NGINX or from behind a proxy? If so, in the meantime, you might be able to just add the access control headers to allow client-side requests from Dashy. Probably something like this:

```
Access-Control-Allow-Origin: https://location-of-dashy/
```

**Reply by EVOTk** [reply]
I am using pihole, install via CURL method
https://docs.pi-hole.net/main/basic-install/

~~If I add:~~
~~`Access-Control-Allow-Origin: https://location-of-dashy/`~~

~~In: /etc/lighttpd/external.conf~~

~~Then the error changes in the browser, and becomes:~~
~~`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://192.168.2.90/admin/api.php. (Reason: CORS request did not succeed). Status code: (null).`~~

**Reply by EVOTk** [reply]
The previous command crashes pi-hole ^^"

If I add:
`setenv.add-response-header = ("Access-Control-Allow-Origin" => "http://192.168.2.44:4000/")`

> http://192.168.2.44:4000/ is my URL of Dashy

So I still have the same error in Dashy:
![2021-12-29 22_15_30-Window](https://user-images.githubusercontent.com/45015615/147704031-46819b83-2212-4e50-be8d-53da1a239473.png)


But the console is different:
```
Stack Trace
Error: Network Error CoolConsole.js: 18: 12
```
![2021-12-29 22_16_28-Window](https://user-images.githubusercontent.com/45015615/147704035-0eee232c-dc05-48e0-af90-c72f93d8ffd9.png)

**Reply by lissy93** [reply]
I'll get it fixed by adding an option to proxy the request, I already wrote the [cors-procy](https://github.com/Lissy93/dashy/blob/master/services/cors-proxy.js) so it should be pretty easy :)
Weird that I wasn't getting that issue on my instance

**Reply by lissy93** [reply]
This is now merged in #392 and released in 1.9.6
You can now Proxy requests by setting `useProxy: true` for any widget, [here's an example](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md#proxying-requests)
Let me know how you get on with that,
Also, 🎇 HAPPY NEW YEAR!! 🎇 :)

**Reply by EVOTk** [reply]
It works perfectly thank you very much!
And happy new year to you and your family / friends :)
![2022-01-01 02_34_18-EVOServ - Dashy — Mozilla Firefox](https://user-images.githubusercontent.com/45015615/147842316-de1a9bf0-af1e-4b99-8536-671170ac57ee.png)
